src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -0,0 +1,189 @@
+              true,
+            )),
+        )
+        .subcommand(
+          SubCommand::with_name("save")
+            .about(
+              "Ingest a file by path, which allows it to be used in Directories/Snapshots. \
+Outputs a fingerprint of its contents.",
+            )
+            .arg(Arg::with_name("path").required(true).takes_value(true)),
+        ),
+    )
+    .subcommand(
+      SubCommand::with_name("directory").subcommand(
+        SubCommand::with_name("save")
+          .about(
+            "Ingest a directory recursively. Saves all files found therein and saves Directory \
+protos for each directory found. Outputs a fingerprint of the canonical top-level Directory proto.",
+          )
+          .arg(Arg::with_name("source").required(true).takes_value(true)),
+      ),
+    )
+    .arg(
+      Arg::with_name("store_dir")
+        .takes_value(true)
+          // TODO: Default this to wherever pants actually stores this.
+        .default_value("/tmp/lmdb"),
+    )
+    .get_matches();
+
+  let store = Store::new(top_match.value_of("store_dir").unwrap()).unwrap();
+
+  match top_match.subcommand() {
+    ("file", Some(sub_match)) => {
+      match sub_match.subcommand() {
+        ("cat", Some(args)) => {
+          let fingerprint = fingerprint_or_die(args.value_of("fingerprint").unwrap());
+          match store.load_bytes(&fingerprint).unwrap() {
+            Some(bytes) => {
+              io::stdout().write(&bytes).unwrap();
+            }
+            None => {
+              die(format!("File with fingerprint {} not found", fingerprint));
+            }
+          }
+        }
+        ("save", Some(args)) => {
+          match save_file(&store, &PathBuf::from(args.value_of("path").unwrap())) {
+            Ok((fingerprint, _)) => println!("{}", fingerprint),
+            Err(err) => die(err),
+          };
+        }
+        (_, _) => unimplemented!(),
+      }
+    }
+    ("directory", Some(sub_match)) => {
+      match sub_match.subcommand() {
+        ("save", Some(args)) => {
+          match save_directory(&store, &PathBuf::from(args.value_of("source").unwrap())) {
+            Ok((fingerprint, _)) => println!("{}", fingerprint),
+            Err(err) => die(err),
+          }
+        }
+        (_, _) => unimplemented!(),
+      }
+    }
+    (_, _) => unimplemented!(),
+  }
+}
+
+fn save_file(store: &Store, path: &Path) -> Result<(Fingerprint, usize), String> {
+  let mut file = File::open(path).map_err(|e| {
+    format!("Error opening file {:?}: {}", path, e.description())
+  })?;
+  let mut buf = Vec::new();
+  file.read_to_end(&mut buf).map_err(|e| {
+    format!("Error reading file {:?}: {}", path, e.description())
+  })?;
+  Ok((store.store_file_bytes(&buf)?, buf.len()))
+}
+
+fn save_directory(store: &Store, root: &Path) -> Result<(Fingerprint, usize), String> {
+  let mut directory = bazel_protos::remote_execution::Directory::new();
+  for maybe_entry in read_dir(root).map_err(|e| {
```

Comment:
It would be good to begin integrating this with the existing code to do this sooner rather than later.

In particular, the existing interface (in pseudocode, ignoring Results) is approximately:
```rust
// Parse globs to include/exclude.
let globs = PathGlobs.create(&["path/to/include/**/*"], &[]);
// Traverse the filesystem in parallel to expand the globs (note: need an impl of VFS; example:
// https://github.com/pantsbuild/pants/blob/01bcda6e3c352d99dc6a583ed57295d3a4c5afe8/src/rust/engine/src/nodes.rs#L73-L92)
let paths: Vec<PathStats> = VFS::expand(globs).wait();
// Capture the PathStats into a Snapshot:
let snapshot = Snapshots::create(.., paths);
```

The first two lines implement filesystem traversal and glob expansion, and it would be good to reuse them here if possible: probably by having the codepath for this utility consume a `Vec<PathStat>` to create the `Directory`, or by changing `VFS::expand` to output something that can be converted more naturally into a recursive structure.

Reply:
Will give it a go now :)